doc/dev/haskell-idioms.md

```diff
@@ -0,0 +1,182 @@
+# Fancy Haskell idioms
+
+This page lists a few idioms that are used in GREASE and might be a bit
+surprising when you first see them. To understand these idioms, you should be
+comfortable with GADTs, kinds, and monads.
+
+## Types with names that start with `Some`
+
+Haskell currently lacks support for [first-class existential types]. They can be
+emulated with [GADT]s like so:
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
```

Comment:
A bit pedantic, but I think this should say that the _return_ type does not mention `x`.